src/memray/_memray.pyx

```diff
@@ -124,18 +124,138 @@ def size_fmt(num, suffix='B'):
+    # 3. Since Python 3.11, one _PyFrame_EvalFrameDefault call can evaluate
+    #    many Python frames. If a frame's is_entry_frame flag is unset, it
+    #    uses the same _PyFrame_EvalFrameDefault call as its caller.
+    # 4. If the interpreter was stripped, we may not be able to recognize
+    #    every (or even any) _PyFrame_EvalFrameDefault call, so we may
+    #    have extra Python frames left after pairing.
+    cdef vector[unsigned char] is_entry_frame
+    native_stack = native_stack_trace(reader, allocator, native_stack_id, generation)
+    python_stack = reader.Py_GetStackFrameAndEntryInfo(python_stack_id, &is_entry_frame)
+
+    cdef ssize_t num_non_entry_frames = count(
+        is_entry_frame.begin(), is_entry_frame.end(), 0
+    )
+
+    # Entry frames replace native frames; non-entry frames are inserted.
+    hybrid_stack = [None] * (len(native_stack) + num_non_entry_frames)
+
+    # Both stacks are from most recent to least, but we must pair things up
+    # least recent to most to handle cases where _PyFrame_EvalFrameDefault
+    # allocated memory before calling the profile function.
+    native_stack.reverse()
+    cdef ssize_t pidx = len(python_stack) - 1
+    cdef ssize_t hidx = len(hybrid_stack) - 1
+
+    cdef ssize_t to_skip = 0
+    if tid == reader.getMainThreadTid():
+        to_skip = reader.getSkippedFramesOnMainThread()
+    cdef ssize_t first_kept_frame = pidx - to_skip
+
+    for native_frame in native_stack:
+        symbol = native_frame[0]
+        if pidx >= 0 and "_PyEval_EvalFrameDefault" in symbol:
+            while True:
+                # If we're not keeping all frames and we've reached the
+                # first one we want to keep, remove frames above it.
+                if to_skip != 0 and pidx == first_kept_frame:
+                    del hybrid_stack[hidx + 1:]
+
+                assert hidx >= 0
+                hybrid_stack[hidx] = python_stack[pidx]
+                hidx -= 1
+                pidx -= 1
+
+                # Stop when we either run out of Python frames or reach the
+                # entry frame being evaluated by the next eval loop.
+                if pidx < 0 or is_entry_frame[pidx]:
+                    break
+        else:
+            assert hidx >= 0
+            hybrid_stack[hidx] = native_frame
+            hidx -= 1
+
+    assert hidx == -1
+    if pidx >= 0:
+        # We ran out of native frames without using up all of our Python
+        # frames. We've seen this happen on stripped interpreters on Alpine
+        # Linux in CI. Presumably this indicates that unwinding failed to
+        # symbolify some of the calls to _PyEval_EvalFrameDefault.
+        return [("<unknown stack>", "<unknown>", 0)]
+
+    return hybrid_stack[:max_stacks]
+
 
 @cython.freelist(1024)
 cdef class AllocationRecord:
     cdef object _tuple
-    cdef object _stack_trace
-    cdef object _native_stack_trace
+    cdef object _stack_trace_cache
```

Comment:
IIRC you may be able to use `cdef dict _stack_trace_cache` and cython will improve access code but I am not 100% sure that the last part is true.

Reply:
Yeah, looks like it actually does generate different code for that. I'll make that change.